Concept: translations (#130)
codewars/docs: content/concepts/kata/translations.md

# Translations

Codewars supports a variety of [programming languages](/languages/) and every kata can have more than one language version available. Users may choose any language available to solve the kata and do so in as many languages as they wish. After gaining the required privileges, users who solved the kata are allowed to create or review translations to make the kata available for a larger audience.

> ## Training on a language version

Reviewer: I just reread and "language version" sounds confusing because it's usually language's version. Not sure what's better though, so you can still merge it whey you're happy with it.

Author: I was thinking "translation" until it gets approved, and when it's merged into a kata, it becomes its "language version". So translation is in forks, and language version is in the kata. But I agree this term is ambiguous and easy to confuse with "language [runtime] version".

I am not sure if we really need such distinction though, so I used "translation" wherever possible.

When the user enters the kata trainer to solve a kata, they can select a language to train on. Some kata are available in many languages, some in just a few, and some only in one or two. If a kata is not available in your favorite language, you must select another language you are familiar with or wait until a translation in your preferred language is approved. Completing a kata in only one language is enough to mark the kata as solved; however, there is something to gain from completing it in other languages too:

- Your [language rank](/concepts/gamification/ranks/#user-rank-breakdown) progresses for every language you solved the kata with (but Honor and overall rank progress are awarded only once per kata).
- You unlock the access to solutions of other users in this language.
- You can view full test suite used to test user attempts against their solutions.

## Creating a translation

To be able to translate a kata to other languages, the user has to complete the kata in at least one language. Once they have completed it, there will be an option within the languages menu to add a new language.

> ![Language Menu](https://www.evernote.com/l/AAW0GaebQllDBb_YS-AfeaUiwq5PoxaDPIoB/image.png)

Author: Done.

## Training on a selected language

When the user enters the kata trainer to solve a kata, they can select a language to train on. Some kata are available in many languages, some in just a few, and some only in one or two. If a kata is not available in your favorite language, you must select another language you are familiar with or wait until a translation in your preferred language is approved. Completing a kata in only one language is enough to mark the kata as solved; however, there is something to gain from completing it in other languages too:

- Your [language rank](/concepts/gamification/ranks/#user-rank-breakdown) progresses for every language you solved the kata with (but Honor and overall rank progress are awarded only once per kata).
- You unlock the access to solutions of other users in this language.
- You can view full test suite used to test user attempts against their solutions.

> ## Creating a translation

Author: You are right, and I will think how to extract it somewhere. But first I'd like to have this branch merged because I think I did something not right when I created it and now it has a veeery long tail of commits from previous branches. I am afraid that if I start moving files here, I will get conflicts which I will not be able to resolve :)

To be able to translate a kata to other languages, the user has to complete the kata in at least one language. Once they have completed it, there will be an option within the languages menu to add a new language.

<div class="block dark:hidden">

![add translation](./img/add-translation_light.png)

</div>
<div class="hidden dark:block">

![add translation](./img/add-translation_dark.png)

</div>

This will take you to the translations page where you can view existing translations and create new ones.

Creating a new translation means writing code snippets required for every kata to have:

- **Example solution**, to show that it's possible to solve the kata.
- **Solution setup**, which works as a stub to work on for users training on the kata.

> - **Sample tests**, with some initial test cases which can be expanded when training on a kata. Sample tests can also serve as an example how the solution is called.

Author: "expanded", as in "new test cases can be added".

> - **Sample tests**, with some initial test cases which can be expanded when training on a kata. Sample tests can also serve as an example how the solution is called.

Reviewer 2: ah ok.

> expanded by the user

then

- such translations cannot be properly reviewed before it's published so it misses the opportunity for bugs and issues to be found by reviewers,
- maintenance of a translation created this way is difficult because it cannot be improved by forking. It can be changed only with the kata editor which is available only to power users.

### Reward

Translators are rewarded for their contribution upon approval of their translation:

- The translator earns rank progress for the individual language in which the translation was authored in, equivalent to solving the kata in that language.
- The translator gains twice the [honor](/references/gamification/honor/#other) compared to solving the kata.
- If the kata is in beta, rank and honor rewards are granted upon approval of the kata itself.
- The translator is listed as a [coauthor (TODO: explain)]() of the kata.

_WARNING: Currently, the co-author flag seems to be affected by a bug which only rewards translators with honor and rank progress if they have not been listed as a co-author of the kata prior to the approval of the translation._

## Reviewing and working with pending translations

> When a translation is published, it's put into a pending state and awaits for review and approval. Sufficiently privileged users can search for them, review them, and request some fixes, approve, or reject them. Currently, pending translations can be found in one of following ways:

Reviewer 2: ```suggestion
When a translation is published, it's put into a pending state and awaits for review and approval. Users who earned the related privileges **(TODO: link to related page)** can search for them, review them and request some fixes through the discourse section of the translation, approve or reject them.
Currently, pending translations can be found in one of following ways:
```

- Reading through the kata discourse and searching for posts where translators announced the fact that they published a translation. Translators are encouraged to post a message labelled as **Suggestion** in the kata discourse with a link and information about the language of the translation, so others could find it easily.

> - Navigating to `https://www.codewars.com/kata/kata-id/translations` page lists all accepted and pending translations of a kata.

Reviewer 2: ```suggestion
- Navigating to `https://www.codewars.com/kata/kata-id/translations` page, that lists all accepted, rejected or pending translations of a kata. Pending translations are at the top of the page, and approved/rejected ones are in the "closed" frame below.
```

Not sure about the addition at the end...

> - A small group of users has been granted the privilege of accessing the page with a list of all pending translations eligible for approval by them. This is the most convenient way of searching for pending translations, but not available to every user at this moment.

Reviewer 2: ```suggestion
- A small group of users has been granted the privilege of accessing the page with a list of all pending translations eligible for approval by them. This is the most convenient way of searching for pending translations, but not available to every user at the moment.
```

### Reviewing a translation

> Before a pending translation is accepted and a new language becomes available for a kata, it needs to be reviewed and determined whether it holds up to Codewars quality standards. When reviewing a translation, users can perform the following steps:

Reviewer 2: ```suggestion
Before a pending translation is accepted and a new language becomes available for a kata, it needs to be reviewed and determined whether it holds up to Codewars quality standards.
When reviewing a translation, users can perform the following steps:
```

- **Reading the code and description**.

> - **Running the tests** against _other_ solutions along with the reference solution. Unfortunately, testing solutions to a proposed translation other than the reference solution is currently rather cumbersome. One possibility is to fork the translation, replace the reference solution with your solution, run the tests and discard the fork afterwards. Another way is to create a new kumite from scratch and copy/paste code snippets from the translation under review to your kumite.

Reviewer 2: >  **Running the tests** against _other_ solutions along with the reference solution.

this sentence isn't clear enough: I know the process and I really have a hard time to figure out what you're talking about, here :s

I guess it's about the order in which you present the information, here. I'd split that in 2 different bullet points

1. running the tests and look at the output (would need a screenshot to show the stuff to click)
2. "if you want to test the code against another solution, ..." -> that's the second idea you're conveying in the current §

Reviewer 2: note: there is that possible bug about: "I want to test the translation, so I open a fork of it and play there with the code, it's ok, so I just go out of the fork without sving/publishing it".
And there... you already lost the hability to gain % in that language, just like if you approve the translation, even if you're not the one who approve it.

not sure this is documented. IIRC, Unnamed posted something about that, but not sure it's in the github repo.

Author: yup, it's not very clear. I will think about it.
I do not know how the bug you mentioned works though.

Reviewer 2: thing is, since it's a bug, I don't really know either... x)

Maybe Unnamed could bring more efficient feedback about that (I didn't approve translation in months if not more)

Reviewer: > note: there is that possible bug about: "I want to test the translation, so I open a fork of it and play there with the code, it's ok, so I just go out of the fork without saving/publishing it".
And there... you already lost the hability to gain % in that language, just like if you approve the translation, even if you're not the one who approve it.

I wasn't aware of this. Is this reported somewhere? It's not obvious how that's possible so I'll need to dig in deeper later.

Reviewer: @error256 Does this happen every time? I was only aware that approving makes you a "coauthor" and that prevents gaining progress.

Reviewer 2: (I just hope my memory is correct about this thing...)
(did I already tell you guys I don't trust my memory more than often...? :/ )

> - **Reading the comments** to check if there are any remarks from other reviewers and whether they still hold or were addressed appropriately. Unfortunately, comments under translations cannot be labelled as an **Issue** or **Suggestion** so there are no additional markers for the severity of the comment.

Reviewer 2: ```suggestion
- **Reading the comments** to check if there are any remarks from other reviewers and whether they still hold or were addressed appropriately. Be aware that comments under translations can neither be labelled as an **Issue** or **Suggestion**,  so there are no additional markers for the severity of the comment, nor be marked as spoilers, so don't tell too much about the solution/code in your comment.
```

- **Verifying** if the translation holds to [translation authoring guidelines (TODO)]().

After a review is completed, the reviewer can finalize it with one of following actions:

- **Leaving a comment** with remarks, suggestions and issues which should be addressed by the translator.

> - **Fixing problems found during the review**. The translation fork may be edited directly only by its author. If a user wants to make changes to a translation fork created by another user, they have to fork it first, edit the code, and publish the changes as a new fork.

Reviewer 2: ~~warning, a reviewer cannot "edit" the translation, he can just create a new fork of it.~~

ok, I let my first comment to show that the order of the sentence needs a change... x)

=> "title. Only the author of the translation ..."

- **Approving the translation**, which introduces a new language to the kata available for solving.
- **Rejecting the translation** if it cannot be fixed.

### Approving a translation

> Pending translations can be approved by either the kata author or any user with the [privilege to coauthor](/references/gamification/privileges/) apart from the translator himself/herself, under the following conditions:

Reviewer 2: ```suggestion
Pending translations can be approved by either the kata author or any user with the [privilege to coauthor](/references/gamification/privileges/). Depending on the people who wants to approve the translation, rules differ a bit:

* **The author** of the kata can of course approve a translation straight away after it's reviewal.
* For **other users**, the following conditions need to met:
  * The approver of course isn't the translator him/herself.
  * The last change to the transltion is older than a week or...
  * The kata author has been inactive for over a month.
```

> - translation is older than a week, or

Reviewer 2: ```suggestion
```
according to the change above

> - kata author has been inactive for over a month.

Reviewer 2: ```suggestion
```
according to the change above

> When the translation is approved, a new language is immediately available for solving. There's also a honor reward, but remember that it's granted to the author of the **initial fork** and not that of the accepted fork!

Reviewer 2: ```suggestion
When the translation is approved, a new language is immediately available for solving. The reward for the translator is triggered too, but remember that it's granted to the author of the **initial fork** and not that of the approved fork!
```

The user who approved the translation should also go through the kata discourse and mark suggestions posted by the translator as resolved.

**NOTE:** at the time of writing, a user who _approves_ a translation is marked as a **kata contributor**. Users with the "contributor" flag seem to be affected by a bug preventing them from earning any future points they would be rewarded for authoring activities on kata they contributed to. Be aware that as a result, approving a translation will render you ineligible for honor points you would otherwise earn by translating, editing or fixing the kata in the future unless the bug is fixed.

### Merge Issues

> Oftentimes if two translations are created around the same time, they may both update the description. When approving the 2nd translation, the 1st approved translation will have already updated the description. This will cause a merge conflict which has to be resolved. You can do this by forking the 2nd translation and updating the description. `<<<<<<<<`, `=========` and `>>>>>>>>` merge labels will show up in the description to help identify where the merge issues are.

Reviewer 2: ```suggestion
Oftentimes if two translations are created around the same time, they may both update the description. When approving the 2nd translation, the 1st approved translation will already have updated the description. This will cause a `merge conflict` which has to be resolved.
You can do this by forking the 2nd translation and updating the description. `<<<<<<<<`, `=========` and `>>>>>>>>` merge labels will show up in the description to help identify where the merge issues are. Another solution is to start a new translation to grab the last version of the kata description, then copy/paste it in the translation to modify.
```

> Once you have fixed the description merge issues you can publish the translation. If you are not the author then you will need to have someone else verify your changes and approve the translation.

Reviewer 2: ```suggestion
Once you have fixed the description merge issues you can publish the translation. If you are not the kata author then you will need to have someone else verify your changes and approve the translation.
```

### Rejecting a translation

> Sometimes the best course of action is to simply reject a pending translation. Some possible reasons could be:

Reviewer 2: ```suggestion
Sometimes the best course of action is to simply reject a pending translation.
Some possible reasons could be:
```

- translation is outdated and kata changed a lot since the translation has been published,
- translation is of bad quality and fixing it is too much work,
- some user authored a newer, better translation and they cannot publish it since there can only be one pending translation per language per kata,
- there's a justified concern that someone might approve a translation of insufficient quality before its issues are fixed.

> It's worth nothing that rejected translations are not wasted work. They are still available in the system, can be forked and serve as a base for new translations, or its code can be simply copied and reused.

Reviewer 2: ```suggestion
It's worth noting that rejected translations are not wasted work. They are still available in the system, can be forked and used as a base for new translations, or its code can be simply copied and reused.
```

## Fixing existing translations

Sometimes a kata needs to be fixed due to an issue, be it a bug in the reference solution, typos in the description or the lack of test coverage. Fixes and improvements to kata can be done in one of two ways:

> - If a translation was created with translations panel, its most recent version can be forked, modified and published, which then goes through the usual translation review process and waits until another user approves it. However, translations created directly within the kata editor cannot be forked.

Reviewer 2: ```suggestion
- If a translation was created through the translations panel, its most recent version can be forked, modified and published, which then goes through the usual translation review process and waits until another user approves it. However, translations created directly within the kata editor cannot be forked.
```

> - A translation may be modified directly within the kata editor. This method of fixing kata is discouraged unless absolutely necessary due to its drawbacks:

Reviewer 2: about that, I tend more and more to use directly the edit panel.

There is sort of a contradiction between having the right ot reach the edit panel and the idea to only go through the translation panel.

Note there is a major drawback to proceed only through translations, even if everyone sticks to the process: when forking a translation while another one has been approved, merge conflicts will apear too... :/

Reviewer 2: WARNING: it seems I accidentally removed a sentence of the file with this comment... :/ That wasnt supposed to happen

Author: Maybe I should not say that this way is discouraged, just that it can cause some other set of problems with outdated forks later.  Will think about it.

Author: > when forking a translation while another one has been approved, merge conflicts will apear too... :/

It's true only for description. By forking, you cannot cause merge conflict for code snippets, but it's possible when code snippet is changed with kata editor. Existing forks become outdated, and when forked again, changes introduced with kata editor are either ignored, or cause merge conflict. 

Maybe this needs some explanation with an example:
 - initial version of translation is created (lets call it `Fork1`), it has fixed test `A`.
 - someone notices that test case `B` is missing. `Fork1` is forked into `Fork2`, test `B` is added.
 - someone notices that test case `C` is missing. They edit the kata with kata editor, and add test case `C`.
 - someone notices that test case `D` is missing. They fork `Fork2`, and... what happens with test case `C`? is it missing from the fork, or is it presented as merge conflict? Does anyone know?

Reviewer 2: Honestly, I'd stay very evasive about all of this in the documentation. The main problem is that there is a structural flaw behind all of this. In the end it's calling for a massive change in the workflow. Too much contradiction everywhere

* access to the edit panel while translation needs a delay => slow the bug fixing. Most users won't do bug fixing if their changes are never reviwed/approved
* going through translations, at each approval, the guy who did shit in the first place get points... XD
* impossible to fix the original language through translations

Fundamentally, there is a contradiction between translation process and maintaining kata.

The flaw is really hue, here. @kazk : any idea/feedback about that?
Thinking about something: changing a bit the UI, allowing to fork a language without regarding the origial translation, and importing by default the current version of the kata. Everything would be up to date, then. It needs more thinking about how to handle rejected/archived translations, though. And what to do about the points too. rewarding a bug fix like closing a parenthese in the deescription? 

do we already have an issue about that somewhere? :o

Author: While I agree it's a flaw in the process, I still think it's good thing to have it documented, even if it's not intended behavior. That's what documentation is for: it answers questions, and explains how things work. It does not have to explain *why* things work this way, but it makes users aware of what they can expect from the system.
I actually received questions about this aspect (not many, but there were two or three people through the time) and I did not know how to answer. When users become able to create translations and kata, they become curious and want explanations. I think docs should give answers to them, even if the answer is "it works so and so, but it's a bug".

Reviewer 2: :+1:

Author: Related issue is https://github.com/codewars/codewars.com/issues/2180 and if you want to discuss it or have some ideas, it would be great.

Reviewer 2: > @hobovsky ([comment](https://github.com/codewars/docs/pull/130#discussion_r514440134)):
> - initial version of translation is created (lets call it Fork1), it has fixed test A.
> -  someone notices that test case B is missing. Fork1 is forked into Fork2, test B is added.
> -  someone notices that test case C is missing. They edit the kata with kata editor, and add test case C.
> -  someone notices that test case D is missing. They fork Fork2, and... what happens with test case C? is it missing from the fork, or is it presented as merge conflict? Does anyone know?

[answer here](https://github.com/codewars/codewars.com/issues/2180#issuecomment-719010602)

- **Tests** which are run against user solution when they consider their solution ready. Tests collected in this snipped must pass to consider the kata solved.
- **Preloaded** snippet, which contains code not visible to users, and availale for both sample tests and full tests.

Description is shared among all language versions, so changing it in your translation will change it for all remaining languages. If you need to add some language specific information in the description, you can use Codewars specific extensions for Markdown formatting: [Sequential Code Blocks](/references/markdown/extensions/#sequential-code-blocks) or [Language Conditional Rendering](/references/markdown/extensions/#conditional-rendering).

> Translation does not have to be finished in one go. It can be saved and kept in draft state as long as necessary. When translator considers it ready, it can be published and put in a "pending" state. It becomes available for other users for review, and can be approved, edited with fixes if necessary, or rejected. Only one translation can be published for given kata and language. For example, if there's already published Java translation, another one can be created but cannot be published until the former one is rejected.

Reviewer 3: ```suggestion
The translation does not have to be finished in one go. It can be saved and kept in draft state as long as necessary. When the translator considers it ready, it can be published which places it in a "pending" state and becomes available for other users to review, approve, edit or reject. Only one translation may be published for any given kata and language. For example, if a published Java translation already exists then a new one may be created but not published until the former is rejected.
```

**NOTE:** kata authors and users with appropriate privileges are able to add new language versions directly with kata editor. This approach is strongly discouraged for following reasons:

- such language version cannot be properly reviewed before it's published, so it misses its opportunity to have bugs and issues found by reviewers,
- maintenance of a language version created this way is difficult, because it cannot by improved by forking. It can be changed only with kata editor, and this way is available only to power users.

### Reward

> Translators are rewarded for their contribution after translation it is approved and becomes available for others to solve:

Reviewer 3: ```suggestion
Translators are rewarded for their contribution upon approval of their translation:
```

> - Rank for the language of the translation increases just as if tranlator solved the kata in this language.

Reviewer 3: ```suggestion
- The translator earns rank progress for the individual language in which the translation was authored in, equivalent to solving the kata in that language.
```

> - Author gains twice as much [Honor points](/references/gamification/honor/#other) as compared to solving the kata.

Reviewer 3: ```suggestion
- The translator gains twice the [honor](/references/gamification/honor/#other) compared to solving the kata.
```

> - If the kata is in beta, rank and Honor rewards are granted once it gets approved and its rank assigned.

Reviewer 3: ```suggestion
- If the kata is in beta, rank and honor rewards are granted upon approval of the kata itself.
```

> - Author is marked as [coauthor (TODO: explain)]() of translated kata.

Reviewer 3: ```suggestion
- The translator is listed as a [coauthor (TODO: explain)]() of the kata.
```

> _*WARNING:* Currently, "Coauthor" flag seems to be affected by a bug which prevents coauthors of a kata from collecting a reward for its translations. Until it gets fixed, users are rewarded for a translation only if they are not a coauthor of the translated kata yet._

Reviewer 3: ```suggestion
_WARNING: Currently, the co-author flag seems to be affected by a bug which only rewards translators with honor and rank progress if they have not been listed as a co-author of the kata prior to the approval of the translation._
```

## Reviewing and working with pending translations

When a translation is published, it's put into a pending state and awaits for review and approval. Sufficiently privileged users can search for them, review them, and request some fixes, approve, or reject them. Currently, pending translations can be found in one of following ways:

- Reading through the kata discourse and searching for posts where translators announced the fact that they published a translation. Translators are encouraged to post a message labelled as **Suggestion** in the kata discourse with a link and information about the language of the translation, so others could find it easily.
- Navigating to `https://www.codewars.com/kata/kata-id/translations` page lists all accepted and pending translations of a kata.
- A small group of users has been granted the privilege of accessing the page with a list of all pending translations eligible for approval by them. This is the most convenient way of searching for pending translations, but not available to every user at this moment.

### Reviewing a translation

> Before pending translation is accepted and new language version becomes available for a kata, it needs to be reviewed and determined whether it keeps up to Codewars quality standards. When reviewing a translation, users can perform following steps:

Reviewer 3: ```suggestion
Before a pending translation is accepted and a new language version becomes available for a kata, it needs to be reviewed and determined whether it holds up to Codewars quality standards. When reviewing a translation, users can perform the following steps:
```

- **Reading the code and description**.

> - **Running tests**, and testing tests by running different solutions. Unfortunately, currently running your own solution to a proposed translation is somewhat cumbersome. One possibility is to fork the translation, put in your own solution in place of proposed one, run tests, and discard the fork afterwards. Another way is to create new kumite from scratch and copy/paste code snippets from reviewed translation to your kumite.

Reviewer 3: ```suggestion
- **Running the tests** against _other_ solutions along with the reference solution. Unfortunately, testing solutions to a proposed translation other than the reference solution is currently rather cumbersome. One possibility is to fork the translation, replace the reference solution with your solution, run the tests and discard the fork afterwards. Another way is to create a new kumite from scratch and copy/paste code snippets from the translation under review to your kumite.
```

> - **Reading comments**, to check if there are any remarks from other reviewers, and whether they still hold and were addressed appropriately. Unfortunately comments under translations cannot be labelled as **Issue** or **Suggestion**, so there are no additional markers for severity of the comment.

Reviewer 3: ```suggestion
- **Reading the comments** to check if there are any remarks from other reviewers and whether they still hold or were addressed appropriately. Unfortunately, comments under translations cannot be labelled as an **Issue** or **Suggestion** so there are no additional markers for the severity of the comment.
```

> - **Verifying** if translation holds to [translation authoring guidelines (TODO)]()

Reviewer 3: ```suggestion
- **Verifying** if the translation holds to [translation authoring guidelines (TODO)]()
```

After a review is completed, the reviewer can finalize it with one of following actions:

- **Leaving a comment** with remarks, suggestions and issues which should be addressed by the translator.

> - **Fixing problems found during review**. Translation fork can be edited directly only by its author. If a user wants to make changes in a translation fork created by another user, they have to fork it first, edit the code, and publish changes as new fork.

Reviewer 3: ```suggestion
- **Fixing problems found during the review**. The translation fork may be edited directly only by its author. If a user wants to make changes to a translation fork created by another user, they have to fork it first, edit the code, and publish the changes as a new fork.
```

> - **Approving the translation**, so new language version gets introduced to the kata and becomes available to solve for other users.

Reviewer 3: ```suggestion
- **Approving the translation**, which introduces a new language version to the kata available for solving.
```

- **Rejecting the translation** if it cannot be fixed.

> ### Aproving a translation

Reviewer 3: ```suggestion
### Approving a translation
```

The translation does not have to be finished in one go. It can be saved and kept in draft state as long as necessary. When the translator considers it ready, it can be published which places it in a "pending" state and becomes available for other users to review, approve, edit or reject. Only one translation may be published for any given kata and language. For example, if a published Java translation already exists then a new one may be created but not published until the former is rejected.

> Unfortunately, currently there's no reliable way to prevent approval of translations of insufficient quality. Unlike beta kata, translations are not a subject to any kind of beta process, and comments in translation discourse cannot be labelled as issues. That's why it's very important to make sure that published translations comply with [translation authoring guidelines (TODO)]().

Reviewer 3: ```suggestion
Unfortunately, there is currently no reliable way to prevent approval of translations of insufficient quality. Unlike beta kata, translations are not subject to any kind of beta process, and comments in the translation discourse cannot be labelled as issues. That's why it's very important to make sure that published translations comply with [translation authoring guidelines (TODO)]().
```

> After translation is published, translator is adviced to post a comment in kata discourse with a link to their translation fork, labelled as **Suggestion**. This will notify other users about the fact and it will be easier for others to find such pending translation.

Reviewer 3: ```suggestion
After a translation is published, the translator is advised to post a comment in the kata discourse with a link to their translation fork labelled as **Suggestion**. This will notify other users about the fact and it will be easier for others to find such pending translations.
```

> **NOTE:** kata authors and users with appropriate privileges are able to add new language versions directly with kata editor. This approach is strongly discouraged for following reasons:

Reviewer 3: - "directly with kata editor" --> "directly within the kata editor"
- "discouraged for following reasons" --> "discouraged for the following reasons"

Reviewer 3: It seems that I missed the opportunity to add suggestion to batch since I only edited the comment to use suggestion blocks after publishing the review. Apologies for the inconvenience caused :sweat_smile:

> - such language version cannot be properly reviewed before it's published, so it misses its opportunity to have bugs and issues found by reviewers,

Reviewer 3: "such language version" --> "such language versions"

> - maintenance of a language version created this way is difficult, because it cannot by improved by forking. It can be changed only with kata editor, and this way is available only to power users.

Reviewer 3: "with kata editor, and this way is available" --> "with the kata editor, which is available"

## Pending translations

> When translation is published, it's put into a pending state and awaits for review and approval. Sufficiently privileged users can search for them, review them, and request some fixes, approve, or reject them. Currently, pending translations can be found in one of following ways:

Reviewer 3: "When translation is published" --> "When a translation is published"

Author: ```suggestion
When a translation is published, it's put into a pending state and awaits for review and approval. Sufficiently privileged users can search for them, review them, and request some fixes, approve, or reject them. Currently, pending translations can be found in one of following ways:
```

> - Reading through kata discourse and searching for posts where translators announced the fact that they published a tranlation. Translators are encouraged to post a message labelled as **Suggestion** in kata discourse with a link and information about the language of the translation, so others could find it easily.

Reviewer 3: - "Reading through kata discourse" --> "Reading through the kata discourse"
- "tranlation" --> "translation"
- "in kata discourse" --> "in the kata discourse"

Author: ```suggestion
- Reading through the kata discourse and searching for posts where translators announced the fact that they published a translation. Translators are encouraged to post a message labelled as **Suggestion** in the kata discourse with a link and information about the language of the translation, so others could find it easily.
```

- Navigating to `https://www.codewars.com/kata/kata-id/translations` page lists all accepted and pending translations of a kata.

> - Small group of users has been granted a privilege of accesing the page with list of all pending translations eligible to approval by them. This is the most convenient way of searching for pending translations, but not available to every user at this moment.

Reviewer 3: - "Small group of users" --> "A small group of users"
- "granted a privilege of" --> "granted the privilege of"
- "accesing" --> "accessing"
- "with list of all pending translations eligible to approval by them" --> "with a list of all pending translations eligible for approval by them"

Author: ```suggestion
- A small group of users has been granted the privilege of accessing the page with a list of all pending translations eligible for approval by them. This is the most convenient way of searching for pending translations, but not available to every user at this moment.
```

> After review is completed, reviewer can finalize it with one of following actions:

Reviewer 3: - "After review is completed" --> "After a review is completed"
- "reviewer can ..." --> "the reviewer can ..."

Author: ```suggestion
After a review is completed, the reviewer can finalize it with one of following actions:
```

### Approving a translation

> Pending translation can be approved either by kata author, or users with [privilege to coauthor](/references/gamification/privileges/) who are not the publisher of the translation fork being approved, under following conditions:

Reviewer 3: ```suggestion
Pending translations can be approved by either the kata author or any user with the [privilege to coauthor](/references/gamification/privileges/) apart from the translator himself/herself, under the following conditions:
```

- translation is older than a week, or
- kata author has been inactive for over a month.

> When translation is approved, new language version is immediately available for other users to solve. There's also a honor reward, but remember that it's granted to the author of the **initial fork**, and not to the author of the fork which got finally accepted!

Reviewer 3: ```suggestion
When the translation is approved, a new language version is immediately available for solving. There's also a honor reward, but remember that it's granted to the author of the **initial fork** and not that of the accepted fork!
```

> Translation approver should also go through the kata discourse and mark suggestions posted by the translator as resolved.

Reviewer 3: ```suggestion
The user who approved the translation should also go through the kata discourse and mark suggestions posted by the translator as resolved.
```

> **NOTE:** currently, probably because of a bug, a user who _approves_ a translation is marked as **kata contributor**. Users with "contributor" flag seem to be affected by a bug which prevents them from getting any future points they would be rewarded for authoring activities on the kata they contributed to. Be aware that as a result, approving of a translation can cause you won't be eligible for any Honor points you would earn by translating, editing or fixing the kata in the future, unless the bug is fixed.

Reviewer 3: ```suggestion
**NOTE:** at the time of writing, a user who _approves_ a translation is marked as a **kata contributor**. Users with the "contributor" flag seem to be affected by a bug preventing them from earning any future points they would be rewarded for authoring activities on kata they contributed to. Be aware that as a result, approving a translation will render you ineligible for honor points you would otherwise earn by translating, editing or fixing the kata in the future unless the bug is fixed.
```

### Merge Issues

Oftentimes if two translations are created around the same time, they may both update the description. When approving the 2nd translation, the 1st approved translation will have already updated the description. This will cause a merge conflict which has to be resolved. You can do this by forking the 2nd translation and updating the description. `<<<<<<<<`, `=========` and `>>>>>>>>` merge labels will show up in the description to help identify where the merge issues are.

Once you have fixed the description merge issues you can publish the translation. If you are not the author then you will need to have someone else verify your changes and approve the translation.

### Rejecting a translation

Sometimes the best course of action is to simply reject a pending translation. Some possible reasons could be:

- translation is outdated and kata changed a lot since the translation has been published,
- translation is of bad quality and fixing it is too much work,

> - some user authored newer, better translation and they cannot publish it because there can be only one pending translation per kata per language,

Reviewer 3: ```suggestion
- some user authored a newer, better translation and they cannot publish it since there can only be one pending translation per language per kata,
```

- there's a justified concern that someone might approve a translation of insufficient quality before its issues are fixed.

> It's worth to note that rejected translations are not wasted work. They are still available in the system, can be forked and serve as a base for new translation, or its code can be simply copied and reused.

Reviewer 3: ```suggestion
It's worth nothing that rejected translations are not wasted work. They are still available in the system, can be forked and serve as a base for new translations, or its code can be simply copied and reused.
```

## Fixing existing language versions

> Sometimes kata needs to be fixed because some issue was found: be it a bug in reference solution, tests, typos in description, or maybe some test cases need to be added. Fixes and improvements to every kata can be done in one of two ways:

Reviewer 3: ```suggestion
Sometimes a kata needs to be fixed due to an issue, be it a bug in the reference solution, typos in the description or the lack of test coverage. Fixes and improvements to kata can be done in one of two ways:
```

> - If language version was created as a translation, its most recent version can be forked, modified, and published. Now it needs to go through normal translation review process and wait until other user approves it. However, language versions which were created directly with kata editor (including initial language version of the kata) cannot be forked.

Reviewer 3: ```suggestion
- If a language version was created as a translation, its most recent version can be forked, modified and published, which then goes through the usual translation review process and waits until another user approves it. However, language versions created directly within the kata editor cannot be forked.
```

> - Language version can be modified directly with kata editor. This way of fixing kata is discouraged unless absolutely necessary, because it has some drawbacks:

Reviewer 3: ```suggestion
- A language version may be modified directly within the kata editor. This method of fixing kata is discouraged unless absolutely necessary due to its drawbacks:
```

>   - it causes merge conflicts for modified snippets if someone would want to publish a fork of this language version some time in the future,

Reviewer 3: ```suggestion
- it causes merge conflicts for modified snippets if someone publishes a fork of this language version sometime in the future,
```

>   - direct edits are not a subject to the review process and other users cannot verify introduced changes.

Reviewer 3: ```suggestion
- direct edits are not subject to the review process and other users cannot verify the introduced changes.
```

Once you have fixed the description merge issues you can publish the translation. If you are not the author then you will need to have someone else verify your changes and approve the translation.

### Rejecting a translation

> ## Fixing existing language version

Reviewer 3: ```suggestion
## Fixing existing language versions
```
